src/cdn.rs

```diff
@@ -117,75 +258,202 @@ impl CdnBackend {
                     .build(),
             )
             .send()
-            .await?;
+            .await?
+            .invalidation()
+            .ok_or_else(|| {
+                anyhow!("missing invalidation information in create-invalidation result")
+            })?
+            .id()
+            .ok_or_else(|| anyhow!("missing invalidation ID in create-invalidation result"))?
+            .to_owned())
+    }
+}
 
-        Ok(())
+#[instrument(skip(conn))]
+pub(crate) fn handle_queued_invalidation_requests(
+    cdn: &CdnBackend,
+    conn: &mut impl postgres::GenericClient,
+    distribution_id: &str,
+) -> Result<()> {
+    info!("handling queued CDN invalidations");
+
+    let active_invalidations = cdn
+        .active_invalidations(distribution_id)
+        .context("error querying active invalidations")?;
+
+    // for now we assume all invalidation paths are wildcard invalidations,
+    // so we apply the wildcard limit.
+    let active_path_invalidations: usize = active_invalidations
+        .iter()
+        .map(|i| i.path_patterns.len())
+        .sum();
+
+    debug!(
+        active_invalidations = active_invalidations.len(),
+        active_path_invalidations, "found active invalidations",
+    );
+
+    // remove the invalidation from the queue when they are completed.
+    // We're only looking at InProgress invalidations,
+    // we don't differentiate between `Completed` ones, and invalidations
+    // missing in the CloudFront `ListInvalidations` response.
+    conn.execute(
+        "DELETE FROM cdn_invalidation_queue
+         WHERE 
+             cdn_distribution_id = $1 AND 
+             created_in_cdn IS NOT NULL AND 
+             NOT (cdn_reference = ANY($2))",
+        &[
+            &distribution_id,
+            &active_invalidations
+                .iter()
+                .map(|i| i.invalidation_id.clone())
+                .collect::<Vec<_>>(),
```

Comment:
It's possible (but unlikely) we get a huge list of active_invalidation from the CDN, say on the order of 10,000. That could make this query big and slow, or worse- it could make the query too big to transmit, and this query would continuously fail, allowing the DB to grow unboundedly (and prevent new invalidations from being queued). It's probably good to apply some reasonable limit to the size of this list.

Reply:
What do you mean with "too big to transmit"? 

You mean postgres query size limits? 

Or limits when fetching the invalidations from CloudFront?